torchtune/datasets/_preference.py

```diff
@@ -0,0 +1,116 @@
+        transform (Optional[Callable]): transform to apply to the sample before formatting to the template.
+            Default is None.
+        column_map (Optional[Dict[str, str]]): a mapping from the expected placeholder names in the template
+            to the column/key names in the sample. If None, assume these are identical.
+        train_on_input (bool): Whether the model is trained on the prompt or not. Default is False.
+        max_seq_len (Optional[int]): Maximum number of tokens in the returned input and label token id lists.
+            Default is None, disabling truncation. We recommend setting this to the highest you can fit in memory
+            and is supported by the model. For example, llama2-7B supports up to 4096 for sequence length.
+        **load_dataset_kwargs (Dict[str, Any]): additional keyword arguments to pass to `load_dataset`.
+    """
+
+    def __init__(
+        self,
+        tokenizer: Tokenizer,
+        source: str,
+        template: InstructTemplate,
+        transform: Optional[Callable] = None,
+        column_map: Optional[Dict[str, str]] = None,
+        max_seq_len: Optional[int] = None,
+        **load_dataset_kwargs: Dict[str, Any],
+    ) -> None:
+        self._tokenizer = tokenizer
+        self._data = load_dataset(source, **load_dataset_kwargs)
+        self.template = template
+        self._transform = transform
+        self._column_map = column_map
+        self.max_seq_len = max_seq_len
+
+    def __len__(self):
+        return len(self._data)
+
+    def __getitem__(self, index: int) -> Dict[str, List[int]]:
+        sample = self._data[index]
+        return self._prepare_sample(sample)
+
+    def _prepare_sample(self, sample: Mapping[str, Any]) -> Dict[str, List[int]]:
+        transformed_sample = self._transform(sample) if self._transform else sample
+        prompt = self.template.format(transformed_sample, self._column_map)
+
+        key_chosen = self._column_map["chosen"]
+        key_rejected = self._column_map['rejected']
+
+        chosen_message = [
+            Message(role="user", content=prompt, masked=True),
+            Message(role="assistant", content=transformed_sample[key_chosen])
+            ]
+        
+        rejected_message = [
+            Message(role="user", content=prompt, masked=True),
+            Message(role="assistant", content=transformed_sample[key_rejected])
+            ]
+        
+        # TODO: Trunction differs from original DPO repo
+        # in DPO: first trunctate prompts, then responses
+        chosen_input_ids, c_masks = self._tokenizer.tokenize_messages(
+            chosen_message, self.max_seq_len
+        )
+        chosen_labels = list(np.where(c_masks, CROSS_ENTROPY_IGNORE_IDX, chosen_input_ids))
+
+        rejected_input_ids, r_masks = self._tokenizer.tokenize_messages(
+            rejected_message, self.max_seq_len
+        )
+        rejected_labels = list(np.where(r_masks, CROSS_ENTROPY_IGNORE_IDX, rejected_input_ids))
```

Comment:
Yes, when both the chosen and rejected messages are lengthy, it might inadvertently truncate most of the rejected parts, which isn't ideal. Do you happen to have any alternative suggestions for this?

Reply:
No this makes sense to me, mainly wanted to clarify my own understanding here.